File: C++/Threads/ThreadCreationPerformance.cpp
#include "AbstractBenchmark.cpp"
#include "Runnable.cpp"
#include "Thread.cpp"

namespace oobench {

    class ThreadCreationPerformance : public Runnable, 
            public AbstractBenchmark {
    public:
        ThreadCreationPerformance() { 
        }

        ThreadCreationPerformance(int argc, const char** argv) 
                : AbstractBenchmark(argc, argv) { 
        }

        virtual int getMajorNumber() const {
            return 6;
        }

        virtual int getMinorNumber() const {
            return 1;
        }

        virtual void run() {
            std::cout << "";
        }

        void test(long items) {
            long count = getCountWithDefault(items);
            std::cout << "*** Testing thread creation performance" 
                    << std::endl;
            
            beginAction(1, "thread creation", count);
            reset();
            ThreadCreationPerformance* performance = 
                    new ThreadCreationPerformance();
            for (long i = 0; i < count; ++i) {
                try {
                    Thread* thread = new Thread(performance);
                    thread->start();
                    thread->join();
                    delete thread;
                    proceed();
                } catch (ThreadException& te)  {
                    std::cerr << "Exception: " << te.message() << std::endl;
                }
            }
            delete performance;
            reset();
            endAction();
        }
    };
}

int main(int argc, const char** argv) {
    SHOW_LOCATION("Threads");
    try {
        oobench::ThreadCreationPerformance performance(argc, argv);
        performance.test(20000L);
    } catch (...) {
        std::cerr << "Unknown exception caught" << std::endl;
    }
    return 0;
    std::exit(0);
}

File: Java/org/oobench/threads/ThreadCreationPerformance.java
package org.oobench.threads;

import org.oobench.common.*;

public class ThreadCreationPerformance extends AbstractBenchmark 
        implements Runnable {
    int maxThreads = 0;
    int id;
    ThreadCreationPerformance parentBench = null;

    ThreadCreationPerformance() {
    }

    ThreadCreationPerformance(int theId, 
            ThreadCreationPerformance theParentBench) {
        id = theId;
        parentBench = theParentBench;
    }

    public int getMajorNumber() {
        return 6;
    }

    public int getMinorNumber() {
        return 1;
    }

     void test() {
        maxThreads = getCountWithDefault(20000);
        for (int c = 0; c < 10; c++) {
            if (c == 9) {
                beginAction(1, "thread creation", maxThreads);
            }
            reset();
            ThreadCreationPerformance performance = 
                    new ThreadCreationPerformance(1, this);
            for (int i = 0; i < maxThreads; i++) {
                Thread thread = new Thread(performance);
                thread.start();
                try {
                    thread.join();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            reset();
            if (c == 9) {
                endAction();
            }
        }
    }

    public static void main(String args[]) {
        showLocation();
        ThreadCreationPerformance bench = new ThreadCreationPerformance();
        bench.setArguments(args);
        bench.test();
    }

    public void run() {
        parentBench.proceed();
    }
}
